rfcs/2022-02-23-9481-config-schema.md

```diff
@@ -0,0 +1,823 @@
+the point of usage.
+
+Additionally, as the metadata is typed, the derive macro can generate rich code that utilizes the
+raw types involved whether than having to deal with downcasted/generalized versions. We utilize this
+in the refrrenced `try_derive_field_default_from_self` function to grab the value of a specific
+field from a value of `Self`, allowing us to continue generating and providing typed metadata as we
+render each field, and that field's fields, and so on.
+
+Additionally, you can see the generated code around things descriptions, where we can layer on
+additional checks to ensure required fields are present, giving us to ability to add in run-time
+checks on top of compile-time checks, furthering extending our goal of misuse resistance.
+
+#### 3. `Configurable` derive macro as a vehicle for enforcing compliance of configuration requirements
+
+Following from the code examples in point 2, we'll explore what the user-defined types would look
+like when using the proposed `Configurable` derive macro:. Some of the fields and other metadata may
+differ from the above example as it would have become too verbose to display above. All of that
+said, let's take a look:
+
+```rust
+/// Configuration for the sink.
+#[derive(Clone)]
+#[configurable_component]
+#[configurable(metadata("status", "beta"))]
+struct SinkConfig {
+    /// The endpoint to send requests to.
+    #[configurable(format(uri), deprecated("url"))]
+    #[serde(alias = "url")]
+    endpoint: String,
+    #[serde(default = default_batch_config_for_sink)]
+    #[configurable(subfield(max_events, range(max = 1000)))]
+    batch: BatchConfig,
+}
+
+/// Controls batching behavior i.e. maximum batch size, the maximum time before a batch is flushed, etc.
+#[derive(Default, Clone)]
+#[configurable_component]
+struct BatchConfig {
+    /// Maximum number of events per batch.
+    #[configurable(non_zero)]
+    max_events: Option<u32>,
+    #[configurable(non_zero)]
+    /// Maximum number of bytes per batch.
+    max_bytes: Option<u32>,
+    #[configurable(non_zero)]
+    /// Maximum period of time a batch can exist before being forcibly flushed.
+    max_timeout: Option<Duration>,
+}
+```
+
+At a high level, the `Configurable` derive macro primarily deals with generating the boilerplate
+`Configurable` implementation for the given type, but goes further with actually being able to
+introspect existing attributes as well as allowing further constraints to be applied. Additionally,
+you can see a distinct attribute macro here -- `configurable_component` -- which we use to apply the
+derive attribute for us.  You can think of `#[configurable_component]` as being a string replacement
+marker for `#[derive(Serialize, Deserialize, Configurable)]`. This lets us enforce that `Serialize`
+and `Deserialize` are derived, along with `Configurable`, as they all must be present for any
+configuration type that we want to include within the schema.
```

Comment:
After considering Jean's comments about trying to hide as many of the implementation details as possible, it struck me that an attribute macro is not only desirable from the perspective of ensuring you can't mistakenly miss a required derive -- `Serialize` and `Deserialize`, in this case -- and generally being shorter, but that it gives us a pathway to introduce future additions to configuration types.

As attribute macros act directly as a replacement mechanism -- replace the attribute itself with whatever the macro spits out for a token stream -- we can easily extend configuration types further in the future in a single place rather than having to go through and modify every callsite.

There's potentially something to be said for hiding the derivations of `Serialize`/`Deserialize`, though.  I don't _believe_ it's an ergonomics issue, nor do I believe it changes the behavior in any way, but it could represent an issue in terms of discoverability.  Not being able to trivially search for instance of `Serialize` or `Deserialize`, etc.